lib/hocon/impl/simple_config_origin.rb

```diff
@@ -25,23 +28,173 @@ def initialize(description, line_number, end_line_number,
     @comments_or_nil = comments_or_nil
   end
 
-  attr_reader :description, :line_number, :end_line_number, :origin_type,
+  attr_reader :_description, :line_number, :end_line_number, :origin_type,
               :url_or_nil, :resource_or_nil, :comments_or_nil
 
 
   def self.new_simple(description)
     self.new(description, -1, -1,
-             Hocon::Impl::OriginType::GENERIC,
+             OriginType::GENERIC,
              nil, nil, nil)
   end
 
   def self.new_file(file_path)
     url = URI.join('file:///', file_path)
     self.new(file_path, -1, -1,
-             Hocon::Impl::OriginType::FILE,
+             OriginType::FILE,
              url, nil, nil)
   end
 
+  def self.new_url(url)
+    u = url.to_external_form
+    SimpleConfigOrigin.new(u, -1, -1, OriginType::URL, u, nil, nil)
+  end
+
+  def self.new_resource(resource, url = nil)
+    desc = nil
+    if url.nil?
+      desc = resource + " @ " + url.to_external_form
+    else
+      desc = resource
+    end
+    SimpleConfigOrigin.new(desc, -1, -1, OriginType::RESOURCE,
+                           url.nil? ? nil : url.to_external_form,
+                           resource, nil)
+  end
+
+  def with_line_number(line_number)
+    if (line_number == @line_number) and
+        (line_number == @end_line_number)
+      self
+    else
+      Hocon::Impl::SimpleConfigOrigin.new(
+          @_description, line_number, line_number,
+          @origin_type, @url_or_nil, @resource_or_nil, @comments_or_nil)
+    end
+  end
+
+  def add_url(url)
+    SimpleConfigOrigin.new(@_description, line_number, end_line_number, origin_type,
+                           url.nil? ? nil : url.to_external_form, resource_or_nil,
+                           comments_or_nil)
+  end
+
+  def with_comments(comments)
+    if Hocon::Impl::ConfigImplUtil.equals_handling_nil?(comments, @comments_or_nil)
+      self
+    else
+      Hocon::Impl::SimpleConfigOrigin.new(
+          @_description, @line_number, @end_line_number,
+          @origin_type, @url_or_nil, @resource_or_nil, comments)
+    end
+  end
+
+  def prepend_comments(comments)
+    if Hocon::Impl::ConfigImplUtil.equals_handling_nil?(comments, @comments_or_nil)
+      self
+    elsif @comments_or_nil.nil?
+      with_comments(comments)
+    else
+      merged = []
```

Comment:
Wondering if we should change this to `merged = comments + @comments_or_nil` for consistency with the below method.